extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -119,6 +126,21 @@ public WindowOperatorQueryFrameProcessor(
     for (int i = 0; i < frameReader.signature().size(); i++) {
       typeStrategies[i] = frameReader.signature().getColumnType(i).get().getNullableStrategy();
     }
+
+    // Get virtual columns to be added to the frame writer.
+    this.partitionBoostVirtualColumn = new SettableLongVirtualColumn(QueryKitUtils.PARTITION_BOOST_COLUMN);
+    final List<VirtualColumn> frameWriterVirtualColumns = new ArrayList<>();
+    if (addVirtualColumns) {
+      frameWriterVirtualColumns.add(partitionBoostVirtualColumn);
+
+      final VirtualColumn segmentGranularityVirtualColumn =
+          QueryKitUtils.makeSegmentGranularityVirtualColumn(jsonMapper, query);
+
+      if (segmentGranularityVirtualColumn != null) {
+        frameWriterVirtualColumns.add(segmentGranularityVirtualColumn);
```

Comment:
Likewise, this needn't be gated behind `addVirtualColumns` flag. Last window would have set the signature correctly (which is why it is getting used), and the other stages wouldn't have used it anyways (otherwise you'd have seen the error earlier).

Reply:
Makes sense, have made the change!